velox/functions/lib/DateTimeFormatter.cpp

```diff
@@ -927,9 +928,113 @@ void parseFromPattern(
 
 } // namespace
 
+// Evaluates the max size of datetime format result to pre-allocate result
+// buffer.
+int32_t DateTimeFormatter::evaluateMaxResultSize(
+    const date::time_zone* timezone) const {
+  int32_t size = 0;
+  for (auto& token : tokens_) {
+    if (token.type == DateTimeToken::Type::kLiteral) {
+      size += token.literal.size();
+      continue;
+    }
+    switch (token.pattern.specifier) {
+      case DateTimeFormatSpecifier::ERA:
+        // Fixed size.
+        size += 2;
+        break;
+      case DateTimeFormatSpecifier::CENTURY_OF_ERA:
+        size += std::max((int)token.pattern.minRepresentDigits, 1);
+        break;
+      case DateTimeFormatSpecifier::YEAR_OF_ERA:
+        // Timestamp is in [-32767-01-01, 32767-12-31] range.
+        size += std::max((int)token.pattern.minRepresentDigits, 6);
+        break;
+      case DateTimeFormatSpecifier::DAY_OF_WEEK_0_BASED:
+      case DateTimeFormatSpecifier::DAY_OF_WEEK_1_BASED:
+        size += std::max((int)token.pattern.minRepresentDigits, 1);
+        break;
+      case DateTimeFormatSpecifier::DAY_OF_WEEK_TEXT:
+        if (token.pattern.minRepresentDigits <= 3) {
+          size += 3;
+        } else {
+          // Max size of element in weekdaysFull.
+          size += 9;
+        }
+        break;
+      case DateTimeFormatSpecifier::YEAR:
+        if (token.pattern.minRepresentDigits == 2) {
+          size += 2;
+        } else {
+          // Timestamp is in [-32767-01-01, 32767-12-31] range.
+          size += std::max((int)token.pattern.minRepresentDigits, 6);
+        }
+        break;
+      case DateTimeFormatSpecifier::DAY_OF_YEAR:
+        size += std::max((int)token.pattern.minRepresentDigits, 3);
+        break;
+      case DateTimeFormatSpecifier::MONTH_OF_YEAR:
+        size += std::max((int)token.pattern.minRepresentDigits, 2);
+        break;
+      case DateTimeFormatSpecifier::MONTH_OF_YEAR_TEXT:
+        if (token.pattern.minRepresentDigits <= 3) {
+          size += 3;
+        } else {
+          // Max size of element in monthsFull.
+          size += 9;
+        }
+        break;
+      case DateTimeFormatSpecifier::DAY_OF_MONTH:
+        size += std::max((int)token.pattern.minRepresentDigits, 2);
+        break;
+      case DateTimeFormatSpecifier::HALFDAY_OF_DAY:
+        // Fixed size.
+        size += 2;
+        break;
+      case DateTimeFormatSpecifier::HOUR_OF_HALFDAY:
+      case DateTimeFormatSpecifier::CLOCK_HOUR_OF_HALFDAY:
+      case DateTimeFormatSpecifier::HOUR_OF_DAY:
+      case DateTimeFormatSpecifier::CLOCK_HOUR_OF_DAY:
+      case DateTimeFormatSpecifier::MINUTE_OF_HOUR:
+      case DateTimeFormatSpecifier::SECOND_OF_MINUTE:
+        size += std::max((int)token.pattern.minRepresentDigits, 2);
```

Comment:
Same with `case DateTimeFormatSpecifier::DAY_OF_MONTH:` and `case DateTimeFormatSpecifier::MONTH_OF_YEAR:`.